nucliadb/nucliadb/ingest/orm/entities.py

```diff
@@ -0,0 +1,261 @@
+#
+# nucliadb is offered under the AGPL v3.0 and as commercial software.
+# For commercial licensing, contact us at info@nuclia.com.
+#
+# AGPL:
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+
+import random
+from typing import AsyncGenerator, List, Optional, Set, Tuple
+
+from nucliadb_protos.knowledgebox_pb2 import (
+    DeletedEntitiesGroups,
+    EntitiesGroup,
+    Entity,
+)
+from nucliadb_protos.nodereader_pb2 import (
+    RelationNodeFilter,
+    RelationPrefixSearchRequest,
+    RelationSearchRequest,
+)
+from nucliadb_protos.noderesources_pb2 import ShardId
+from nucliadb_protos.nodewriter_pb2 import SetGraph
+from nucliadb_protos.utils_pb2 import JoinGraph, JoinGraphEdge, Relation, RelationNode
+from nucliadb_protos.writer_pb2 import GetEntitiesGroupResponse, GetEntitiesResponse
+
+from nucliadb.ingest.maindb.driver import Transaction
+from nucliadb.ingest.maindb.keys import (
+    KB_DELETED_ENTITIES_GROUPS,
+    KB_ENTITIES,
+    KB_ENTITIES_GROUP,
+)
+from nucliadb.ingest.orm import knowledgebox
+
+
+class EntitiesManager:
+    def __init__(self, knowledgebox: "knowledgebox.Knowledgebox", txn):
+        self.kb = knowledgebox
+        self.txn = txn
+        self.kbid = self.kb.kbid
+
+    async def get_entities(self, entities: GetEntitiesResponse):
+        async for group, eg in self.entitiesgroups_iterator(exclude_deleted=True):
+            entities.groups[group].CopyFrom(eg)
+
+    async def get_entitiesgroup(
+        self, group: str, entitiesgroup: GetEntitiesGroupResponse
+    ):
+        deleted_groups = await self.get_deleted_entitiesgroups()
+        if group in deleted_groups:
+            return
+        eg = await self.get_entitiesgroup_inner(group)
+        if eg is not None:
+            entitiesgroup.group.CopyFrom(eg)
+
+    async def set_entities(self, group: str, entities: EntitiesGroup):
+        new_entities = EntitiesGroup()
+        stored_entities = await self.get_entitiesgroup_inner(group)
+
+        if stored_entities is None:
+            new_entities = entities
+        else:
+            merged = {}
+
+            for key, entity in stored_entities.entities.items():
+                if key not in entities.entities:
+                    entity.status = Entity.DiffStatus.DELETED
+                else:
+                    entity.status = entities.entities[key].status
```

Comment:
mh, so we are delegating the responsability to decide the status (MERGED vs NORMAL) to the caller of the endpoint? I wonder if it would be simpler to do it here by comparing if the new one is equal to what we have stored.

Reply:
Merged in the entities context means two relations mean the same. Maybe it's not the best word to describe it. For example, I can have "house cat" and "domestic cat". In the UI I choose "domestic cat" to be the same as "house cat" and it'll set the following:
```
{
    "house cat": {"value": "house cat", "represents": ["domestic cat"], "merged": false},
    "domestic cat": {"value": "domestic cat", "represents": [], "merged": true},
}
```